File: algorithms/cpp/17_letter_combinations_of_a_phone_number/PhoneNumber.cc
class PhoneNumber {
// 1. iterative
public:
	vector<string> letterCombinations(string digits) {
        if (digits.empty()) {
            return {};
        }
        vector<string> res;
        res.push_back("");
        
        for (char digit : digits) {
            vector<string> tmp;
            for (char c : book[digit - '0']) {
                for (string s : res) {
                    tmp.push_back(s + c);
                }
            }
            res.swap(tmp);
        }
        return res;
    }
private:
	const vector<string> book = {
        "", "", "abc", "def",
         "ghi", "jkl", "mno",
        "pqrs", "tuv", "wxyz"
    };

// 2. backtracking
public:
    vector<string> letterCombinations1(string digits) {
        if (digits.empty()) {
            return {};
        }
        vector<string> res;
        dfs(digits, "", res, 0);
        return res;
    }
private:
    void dfs(string& digits, string path, vector<string>& res, int index) {
        if (index == digits.size()) {
            res.push_back(path);
            return;
        }
        int cur = digits[index] - '0';
        for (char c : book[cur]) {
            dfs(digits, path + c, res, index + 1);
        }
    }
    const vector<string> book = {
        "", "", "abc", "def",
         "ghi", "jkl", "mno",
        "pqrs", "tuv", "wxyz"
    };
};

File: algorithms/java/17_letter_combinations_of_a_phone_number/PhoneNumber.java
class PhoneNumber {
	public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        if (digits == null || digits.length() == 0) {
            return res;
        }
        List<char[]> map = buildMap();
        dfs(digits, 0, res, map, new StringBuilder());
        return res;
    }
    
    private void dfs(String digits, int index, List<String> res, List<char[]> map, StringBuilder sb) {
        if (index == digits.length()) {
            res.add(sb.toString());
            return;
        }
        int cur = digits.charAt(index) - '0';
        char[] candidates = map.get(cur);
        for (char c : candidates) {
            sb.append(c);
            dfs(digits, index + 1, res, map, sb);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
    
    private List<char[]> buildMap() {
        List<char[]> res = new ArrayList<>();
        res.add(new char[0]);
        res.add(new char[0]);
        res.add(new char[]{'a', 'b', 'c'});
        res.add(new char[]{'d', 'e', 'f'});
        res.add(new char[]{'g', 'h', 'i'});
        res.add(new char[]{'j', 'k', 'l'});
        res.add(new char[]{'m', 'n', 'o'});
        res.add(new char[]{'p', 'q', 'r', 's'});
        res.add(new char[]{'t', 'u', 'v'});
        res.add(new char[]{'w', 'x', 'y', 'z'});
        return res;
    }
}
